homeassistant/components/forked_daapd/media_player.py

```diff
@@ -0,0 +1,673 @@
+
+    async def async_play_media(self, media_type, media_id, **kwargs):
+        """Play a URI."""
+        if media_type == MEDIA_TYPE_MUSIC:
+            saved_state = self.state  # save play state
+            await self.async_turn_off()  # pauses and saves output states
+            await self._set_tts_volumes()
+            await asyncio.sleep(self._tts_pause_time)
+            # save position
+            saved_song_position = self._player["item_progress_ms"]
+            saved_queue = (
+                self._queue if self._queue["count"] > 0 else None
+            )  # stash queue
+            if saved_queue:
+                saved_queue_position = next(
+                    i
+                    for i, item in enumerate(saved_queue["items"])
+                    if item["id"] == self._player["item_id"]
+                )
+                await self._api.clear_queue()
+            self._tts_requested = True
+            await self._api.add_to_queue(uris=media_id, playback="start")
+            try:
+                await asyncio.wait_for(
+                    self._tts_playing_event.wait(), timeout=TTS_TIMEOUT
+                )
+                # we have started TTS, now wait for completion
+                await asyncio.sleep(
+                    self._queue["items"][0]["length_ms"]
+                    / 1000  # player may not have updated yet so grab length from queue
+                    + self._tts_pause_time
+                )
+            except asyncio.TimeoutError:
+                self._tts_requested = False
+                _LOGGER.warning("TTS request timed out.")
+            self._tts_playing_event.clear()
+            # TTS done, return to normal
+            await self.async_turn_on()  # restores outputs
+            await self._api.clear_queue()
+            if self._pipe_control_api:  # resume pipe
+                if saved_state == STATE_PLAYING:
+                    await self.async_media_play()
+            else:  # restore stashed queue
+                if saved_queue:
+                    uris = ""
+                    for item in saved_queue["items"]:
+                        uris += item["uri"] + ","
+                    await self._api.add_to_queue(
+                        uris=uris,
+                        playback="start",
+                        playback_from_position=saved_queue_position,
+                    )
+                    await self._api.seek(position_ms=saved_song_position)
+                    if saved_state != STATE_PLAYING:
+                        await self.async_media_pause()
+        else:
+            _LOGGER.debug("Media type '%s' not supported.", media_type)
+
+
+class ForkedDaapdUpdater:
+    """Manage updates for the forked-daapd device."""
+
+    def __init__(self, hass, api):
+        """Initialize."""
+        self.hass = hass
+        self._api = api
+        self.websocket_handler = None
+        self._all_output_ids = []
```

Comment:
If output ids are unique we should use a set to store them.

Reply:
Yeah, we can use sets.